templates/tower-project.yaml

```diff
@@ -0,0 +1,454 @@
+                  - logs:Describe*
+                  - logs:FilterLogEvents
+                  - logs:Get*
+                  - logs:List*
+                  - logs:StartQuery
+                  - logs:StopQuery
+                  - logs:TestMetricFilter
+                Resource: "*"
+
+  TowerForgePassRolePolicy:
+    Type: AWS::IAM::Policy
+    Properties: 
+      PolicyName: "TowerForgePassRolePolicy"
+      PolicyDocument: 
+        Version: "2012-10-17"
+        Statement: 
+          - Effect: "Allow"
+            Action:
+              - "iam:GetRole"
+              - "iam:PassRole"
+            Resource: 
+              - !GetAtt TowerForgeInstanceRole.Arn
+      Roles: 
+        - !Ref TowerForgeServiceRole
+      Users: 
+        - !Ref TowerForgeServiceUser
+
+  EncryptionKeyStack:
+    Type: AWS::CloudFormation::Stack
+    Properties:
+      TemplateURL: !Sub ${TemplateRootUrl}/aws-infra/v0.3.1/KMS/kms-key.yaml
+      TimeoutInMinutes: 5
+      Parameters:
+        AliasName: !Sub "alias/${AWS::StackName}-EncryptionKey"
+        AdminPrincipalArns: !Join
+          - ","
+          - - !Sub "arn:aws:iam::${AWS::AccountId}:root"
+            - !Join
+              - ","
+              - !Ref AdminArns
+        UserPrincipalArns: !Join
+          - ","
+          - - !Sub "arn:aws:iam::${AWS::AccountId}:root"
+            - !GetAtt TowerForgeServiceRole.Arn
+            - !GetAtt TowerForgeInstanceRole.Arn
+            - !Join
+              - ","
+              - !Ref AdminArns
+            - !If
+              - HasReadWriteAccessArns
+              - !Join
+                - ","
+                - !Ref ReadWriteAccessArns
+              - !Ref AWS::NoValue
+            - !If
+              - HasReadOnlyAccessArns
+              - !Join
+                - ","
+                - !Ref ReadOnlyAccessArns
+              - !Ref AWS::NoValue
+
+  TowerBucket:
+    Type: AWS::S3::Bucket
+    Properties:
+      BucketName: !Sub "${AWS::StackName}-tower-bucket"
+      OwnershipControls:
+        Rules:
+          - ObjectOwnership: BucketOwnerPreferred
+      BucketEncryption:
+        ServerSideEncryptionConfiguration:
+          - BucketKeyEnabled: true
+            ServerSideEncryptionByDefault:
+              SSEAlgorithm: aws:kms
+              KMSMasterKeyID: !GetAtt EncryptionKeyStack.Outputs.KeyArn
+      LifecycleConfiguration:
+        Rules:
+        - Id: DataLifecycleRule
+          Status: !Ref EnableDataLifeCycle
+          ExpirationInDays: !Ref LifecycleDataExpiration
+          Transitions:
+            - TransitionInDays: !Ref LifecycleDataTransition
+              StorageClass: !Ref LifecycleDataStorageClass
+
+  TowerBucketPolicy:
+    Type: AWS::S3::BucketPolicy
+    Properties:
+      Bucket: !Ref TowerBucket
+      PolicyDocument:
+        Version: "2012-10-17"
+        Statement:
+          - Sid: DenyAccessToNonAdmins
+            Effect: Deny
+            Principal:
+              AWS: !Ref NonAdminArns
```

Comment:
Because the SSO `Viewer` role is within the account and is bound to the AWS-managed `ViewOnlyAccess` policy, anyone assuming that role will have read-access to the project-specific buckets if we don't include this Deny statement (if I'm not mistaken). 

I would like to get rid of the `AccountNonAdminArns` (formerly `NonAdminArns`) if possible. We could add a Deny for `s3:*` on the SSO `Viewer` role and expect users to use their sandbox `Developer` role to access their project bucket. My only hesitation is whether blocking all S3 access on the `Viewer` role will make debugging more difficult. I can't think of anything, and we could pursue this approach now and change it later if we do find a fatal flaw.

Reply:
As discussed in phone call, I prefer to revisit what permissions the Viewer role has rather than using a Deny. Let's opt-in to a narrower range of permissions.

Looking it up, the Viewer has AWS managed policy `ViewOnlyAccess` which includes `s3:ListAllMyBuckets` and `s3:ListBucket`. 
ListBucket allows seeing the objects within a bucket, but not pulling those.
The misleadingly named ListAllMyBuckets allows listing of all buckets, according this [knowledge center article](https://aws.amazon.com/premiumsupport/knowledge-center/s3-console-access-certain-bucket/).

Since ViewOnlyAccess allows access to many things that don't seem necessary, let's craft our own, narrower policy that does not allow our Viewer role to see things in S3. But we don't need to use a Deny. We can explicitly opt-in as @zaro0508 suggests.

Reply:
Done in Sage-Bionetworks-IT/organizations-infra#282